metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -350,12 +350,6 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
         Uuid prevIncarnationId = null;
         if (existing != null) {
             prevIncarnationId = existing.incarnationId();
-            if (heartbeatManager.hasValidSession(brokerId, existing.epoch())) {
-                if (!request.incarnationId().equals(prevIncarnationId)) {
-                    throw new DuplicateBrokerRegistrationException("Another broker is " +
-                        "registered with that broker id.");
-                }
-            }
```

Comment:
we fence broker registrations for the same ID while the session is active, this ensures that a broker goes through the fenced -> re-registered transition.